### PR TITLE
fix: plastic/elastic slip fix for EFEM

### PR DESCRIPTION
**Pull request by sohailwaziri**
update/store elastic slip only once at the end of the time step

### CODE REVIEW

File: src/coreComponents/physicsSolvers/contact/SolidMechanicsEmbeddedFractures.cpp [hunk 1]
```diff
@@ -767,7 +769,9 @@ bool SolidMechanicsEmbeddedFractures::updateConfiguration( DomainPartition & dom
     {
       arrayView1d< integer const > const & ghostRank = subRegion.ghostRank();
       arrayView2d< real64 const > const & dispJump = subRegion.getField< fields::contact::dispJump >();
+      arrayView2d< real64 const > const & oldJump = subRegion.getField< fields::contact::oldDispJump >();
       arrayView2d< real64 const > const & traction = subRegion.getField< fields::contact::traction >();
+      arrayView1d< real64 const > const & pressure = subRegion.template getField< fields::flow::pressure >();
```
Comment (paveltomin) [inline src/coreComponents/physicsSolvers/contact/SolidMechanicsEmbeddedFractures.cpp]: should not be here...

File: src/coreComponents/constitutive/contact/CoulombFriction.hpp [hunk 2]
```diff
@@ -269,7 +280,7 @@ inline void CoulombFrictionUpdates::computeShearTraction( localIndex const k,
       dTractionVector_dJump[2][2] = m_shearStiffness;
 
       // The slip is only elastic: we add the full slip to the elastic one
-      LvArray::tensorOps::add< 2 >( m_elasticSlip[k], slip );
+      //LvArray::tensorOps::add< 2 >( m_elasticSlip[k], slip );
```
Comment (paveltomin) [inline src/coreComponents/constitutive/contact/CoulombFriction.hpp]: @CusiniM this update and update on the lines 314-315 below got moved into `updateFractureState`.
The reason is: those quantities should not be updated at every Newton but once at the end of the time step insted.
`updateFractureState` is probably not the best place either, what would be the better one?

File: src/coreComponents/physicsSolvers/multiphysics/poromechanicsKernels/ThermalSinglePhasePoromechanicsEFEM_impl.hpp [hunk 3]
```diff
@@ -175,8 +175,8 @@ complete( localIndex const k,
 
   localIndex const embSurfIndex = m_cellsToEmbeddedSurfaces[k][0];
   // Energy balance accumulation
-  real64 const volume        =  m_elementVolume( embSurfIndex ) + m_deltaVolume( embSurfIndex );
-  real64 const volume_n      =  m_elementVolume( embSurfIndex );
+  real64 const volume        =  m_elementVolumeCell( embSurfIndex ) + m_deltaVolume( embSurfIndex );
```
Comment (paveltomin) [inline src/coreComponents/physicsSolvers/multiphysics/poromechanicsKernels/ThermalSinglePhasePoromechanicsEFEM_impl.hpp]: @Guotong-Ren @sohailwaziri should be "frac" here?

File: src/coreComponents/constitutive/contact/CoulombFriction.hpp (hunk 2 above)
Comment (paveltomin) [reply src/coreComponents/constitutive/contact/CoulombFriction.hpp]: @CusiniM could you please confirm that you agree with this change?

**Comment by paveltomin** [comment]
@CusiniM and @Guotong-Ren please have a look and double-check if anything is wrong or missing in this PR

File: src/coreComponents/constitutive/contact/CoulombFriction.hpp (hunk 2 above)
Comment (CusiniM) [reply src/coreComponents/constitutive/contact/CoulombFriction.hpp]: That should be okay.

File: src/coreComponents/constitutive/contact/CoulombFriction.hpp [hunk 4]
```diff
@@ -322,6 +333,22 @@ inline void CoulombFrictionUpdates::updateFractureState( localIndex const k,
...
+                             dispJump[2] - oldDispJump[2] };
+    if( yield < 0 )
+    {
+      // The slip is only elastic: we add the full slip to the elastic one
+      LvArray::tensorOps::add< 2 >( m_elasticSlip[k], slip );
+    }
+    else
+    {
+      m_plasticSlip[k][0] += tractionVector[1] / m_shearStiffness;
+      m_plasticSlip[k][1] += tractionVector[2] / m_shearStiffness;
```
Comment (CusiniM) [inline src/coreComponents/constitutive/contact/CoulombFriction.hpp]: I think that this was wrong and this is the elastic slip. 

Screenshot 2024-11-21 at 17 28 19

File: src/coreComponents/constitutive/contact/CoulombFriction.hpp [hunk 5]
```diff
@@ -322,6 +333,22 @@ inline void CoulombFrictionUpdates::updateFractureState( localIndex const k,
...
+
+      LvArray::tensorOps::copy< 2 >( m_elasticSlip[k], slip );
+      LvArray::tensorOps::subtract< 2 >( m_plasticSlip[k], m_plasticSlip[k] );
```
Comment (CusiniM) [inline src/coreComponents/constitutive/contact/CoulombFriction.hpp]:
```suggestion
      LvArray::tensorOps::subtract< 2 >( m_elasticSlip[k], m_plasticSlip[k] );
```
?

File: src/coreComponents/constitutive/contact/CoulombFriction.hpp (hunk 4 above)
Comment (CusiniM) [reply src/coreComponents/constitutive/contact/CoulombFriction.hpp]: and we only need to keep track of one of the two.

File: src/coreComponents/physicsSolvers/contact/SolidMechanicsLagrangeContact.cpp [hunk 6]
```diff
@@ -2262,7 +2262,8 @@ bool SolidMechanicsLagrangeContact::updateConfiguration( DomainPartition & domai
             integer const originalFractureState = fractureState[kfe];
             if( originalFractureState == FractureState::Open )
             {
-              if( dispJump[kfe][0] <= -normalDisplacementTolerance[kfe] )
+              // trying comparison in absolute
+              if( LvArray::math::abs( dispJump[kfe][0] ) <= LvArray::math::abs( normalDisplacementTolerance[kfe] ) )
```
Comment (CusiniM) [inline src/coreComponents/physicsSolvers/contact/SolidMechanicsLagrangeContact.cpp]:
```suggestion
              if( LvArray::math::abs( dispJump[kfe][0] ) <= normalDisplacementTolerance[kfe]  )
```
I think the tolerance is always positive

File: src/coreComponents/constitutive/contact/CoulombFriction.hpp (hunk 4 above)
Comment (paveltomin) [reply src/coreComponents/constitutive/contact/CoulombFriction.hpp]: done

File: src/coreComponents/constitutive/contact/CoulombFriction.hpp (hunk 5 above)
Comment (paveltomin) [reply src/coreComponents/constitutive/contact/CoulombFriction.hpp]: done

File: src/coreComponents/physicsSolvers/contact/SolidMechanicsEmbeddedFractures.cpp [hunk 7]
```diff
@@ -159,6 +159,25 @@ void SolidMechanicsEmbeddedFractures::implicitStepComplete( real64 const & time_
...
+    FrictionBase const & frictionLaw = getConstitutiveModel< FrictionBase >( subRegion, frictionLawName );
+    arrayView1d< integer const > const & ghostRank = subRegion.ghostRank();
+    arrayView2d< real64 const > const & traction = subRegion.getField< fields::contact::traction >();
+    arrayView1d< integer > const & fractureState = subRegion.getField< fields::contact::fractureState >();
+    constitutiveUpdatePassThru( frictionLaw, [&] ( auto & castedFrictionLaw )
+    {
+      using FrictionType = TYPEOFREF( castedFrictionLaw );
+      typename FrictionType::KernelWrapper frictionWrapper = castedFrictionLaw.createKernelUpdates();
+
+      forAll< parallelHostPolicy >( subRegion.size(), [=] ( localIndex const kfe )
```
Comment (CusiniM) [inline src/coreComponents/physicsSolvers/contact/SolidMechanicsEmbeddedFractures.cpp]:
```suggestion
      forAll< parallelDevicePolicy<> >( subRegion.size(), [=] GEOS_HOST_DEVICE ( localIndex const kfe )
```

**Review by CusiniM — APPROVED** [review]

**Pull request merged**